python-checks/src/main/java/org/sonar/python/checks/ConsistentReturnCheck.java

```diff
@@ -0,0 +1,98 @@
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.python.checks;
+
+import java.util.List;
+import java.util.stream.Collectors;
+import org.sonar.check.Rule;
+import org.sonar.plugins.python.api.PythonSubscriptionCheck;
+import org.sonar.plugins.python.api.SubscriptionContext;
+import org.sonar.plugins.python.api.cfg.CfgBranchingBlock;
+import org.sonar.plugins.python.api.cfg.ControlFlowGraph;
+import org.sonar.plugins.python.api.tree.FunctionDef;
+import org.sonar.plugins.python.api.tree.ReturnStatement;
+import org.sonar.plugins.python.api.tree.Statement;
+import org.sonar.plugins.python.api.tree.Tree;
+import org.sonar.plugins.python.api.tree.Tree.Kind;
+import org.sonar.plugins.python.api.tree.WhileStatement;
+
+@Rule(key = "S3801")
+public class ConsistentReturnCheck extends PythonSubscriptionCheck {
+
+  public static final String MESSAGE = "Refactor this function to use \"return\" consistently.";
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Kind.FUNCDEF, ctx -> {
+      FunctionDef functionDef = (FunctionDef) ctx.syntaxNode();
+      ControlFlowGraph cfg = ControlFlowGraph.build(functionDef, ctx.pythonFile());
+      if (cfg == null || hasExceptOrFinally(cfg)) {
+        return;
+      }
+      List<Tree> endStatements = cfg.end().predecessors().stream()
+        .map(block -> parentStatement(block.elements().get(block.elements().size() - 1)))
+        .filter(s -> !s.is(Kind.RAISE_STMT, Kind.ASSERT_STMT, Kind.WITH_STMT) && !isWhileTrue(s))
+        .collect(Collectors.toList());
+
+      List<Tree> returnsWithValue = endStatements.stream()
+        .filter(s -> s.is(Kind.RETURN_STMT) && hasValue((ReturnStatement) s))
+        .collect(Collectors.toList());
+
+      if (returnsWithValue.size() != endStatements.size() && !returnsWithValue.isEmpty()) {
+        addIssue(ctx, functionDef, endStatements);
+      }
+    });
+  }
+
+  private static boolean hasExceptOrFinally(ControlFlowGraph cfg) {
+    return cfg.blocks().stream().anyMatch(block ->
+      block instanceof CfgBranchingBlock && ((CfgBranchingBlock) block).branchingTree().is(Kind.EXCEPT_CLAUSE, Kind.FINALLY_CLAUSE));
+  }
+
+  private static boolean isWhileTrue(Statement statement) {
+    return statement.is(Kind.WHILE_STMT) && Expressions.isTruthy(((WhileStatement) statement).condition());
+  }
+
+  private static void addIssue(SubscriptionContext ctx, FunctionDef functionDef, List<Tree> endStatements) {
+    PreciseIssue issue = ctx.addIssue(functionDef.name(), MESSAGE);
+    for (Tree statement : endStatements) {
+      if (statement.is(Kind.RETURN_STMT)) {
+        ReturnStatement returnStatement = (ReturnStatement) statement;
+        boolean hasValue = hasValue(returnStatement);
+        issue.secondary(statement, String.format("Return %s value", hasValue ? "with" : "without"));
+      } else if (statement.is(Kind.IF_STMT, Kind.FOR_STMT, Kind.WHILE_STMT)) {
```

Comment:
I think we should highlight only the first token for all compound statements, i.e. also for match/case statements and function / class definitions (although it's unlike to end a function with those).